src/daemon.py

```diff
@@ -9,43 +9,92 @@
 
 
-def get_finalized_block_number(web3: Web3) -> int:
+def get_start_block(
+    chain_name: str, solver_slippage_connection: Engine, web3: Web3
+) -> int:
     """
-    Get the number of the most recent finalized block.
+    Retrieve the most recent block already present in raw_token_imbalances table,
+    delete entries for that block, and return this block number as start_block.
+    If no entries are present, fallback to get_finalized_block_number().
     """
-    return web3.eth.block_number - 67
+    try:
+        solver_slippage_connection = check_db_connection(
+            solver_slippage_connection, "solver_slippage"
+        )
+        query_max_block = read_sql_file("src/sql/select_max_block.sql")
+
+        with solver_slippage_connection.connect() as connection:
+            result = connection.execute(
+                text(query_max_block), {"chain_name": chain_name}
+            )
+            row = result.fetchone()
+            max_block = (
+                row[0] if row is not None else None
+            )  # Fetch the maximum block number
+            if max_block is not None:
+                logger.debug("Fetched max block number from database: %d", max_block)
+
+            # If no entries present, fallback to get_finalized_block_number()
+            if max_block is None:
+                return get_finalized_block_number(web3)
+
+            # delete entries for the max block from the table
+            delete_sql = read_sql_file("src/sql/delete_entries_max_block.sql")
+
+            try:
+                connection.execute(
+                    text(delete_sql),
+                    {"chain_name": chain_name, "block_number": max_block},
+                )
+                connection.commit()
+                logger.debug(
+                    "Successfully deleted entries for block number: %s", max_block
+                )
+            except Exception as e:
+                logger.error(
+                    "Failed to delete entries for block number %s: %s", max_block, e
+                )
+
+            return max_block
+    except Exception as e:
+        logger.error("Error accessing database: %s", e)
+        return get_finalized_block_number(web3)
 
 
 def fetch_tx_data(
     backend_db_connection: Engine, start_block: int, end_block: int
 ) -> List[Tuple[str, int, int]]:
-    """Fetch transaction hashes beginning from start_block to end_block."""
+    """
+    Fetch transaction data beginning from start_block to end_block.
+    Returns (tx_hash, auction_id, block_number) as a tuple.
+    """
 
     backend_db_connection = check_db_connection(backend_db_connection, "backend")
+    query = read_sql_file("src/sql/select_transactions.sql")
 
-    query = f"""
-    SELECT tx_hash, auction_id, block_number
-    FROM settlements
-    WHERE block_number >= {start_block}
-    AND block_number <= {end_block}
-    """
-    db_data = pd.read_sql(query, backend_db_connection)
+    db_data = pd.read_sql(
```

Comment:
I guess it was a conscious decision to use pandas here but not in the `get_start_block()` function, right? I don't mind, just wanted to confirm that using 2 different ways of querying the db was done for a reason

Reply:
Since our first test last month was just fetching data from the backend, I had used Felix's ebbo or circuit breaker code which used pandas. I then saw a way without pd and stuck to that. The latest code has this function removed though.